[JIT] Remove Dropout during Frozon Optimization

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51589 [JIT] Remove Dropout during Frozon Optimization**

Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

Commit:
[JIT] Remove Dropout during Frozon Optimization

Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[JIT] Remove Dropout during Frozon Optimization"


Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

[ghstack-poisoned]

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
+
+        mod = torch.jit.script(Net())
+        # set optimize to False here, by default freezing runs optimize_frozen_module
+        frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()), optimize=False)
```
While we're both pretty sure that it removing dropout doesn't effect results, it might be good to test it by comparing the results of the eager module with the frozen one after optimization

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.conv1 = nn.Conv2d(1, 32, 3, 1)
+                self.conv2 = nn.Conv2d(32, 64, 3, 1)
```
Nit: the conv and linear here are quite big - they could be smaller without effecting the test and it would run quicker (if we run the model...). I would also consider making the module simpler / remove some layers to parse the test easier.

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
+        mod = torch.jit.script(Net())
+        # set optimize to False here, by default freezing runs optimize_frozen_module
+        frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()), optimize=False)
+        # inspect frozen mod
```
We should check that "aten::dropout" and "aten::feature_dropout" exist in the graph before our optimization so we know we're actually removing them

eellison commented on torch/csrc/jit/passes/remove_dropout.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
```
I think this isn't being used

eellison commented on torch/csrc/jit/passes/frozen_graph_optimizations.cpp:
```diff
@@ -14,6 +15,7 @@ void OptimizeFrozenGraph(std::shared_ptr<Graph>& graph) {
     FoldFrozenConvBatchnorm(graph);
     FoldFrozenConvAddOrSub(graph);
     FoldFrozenConvMulOrDiv(graph);
+    removeDropout(graph);
```
Nit: the other passes in this are run multiple times to catch different patterns of Conv-BatchNorm-Add / Conv-Add-Mul / Conv-Mul-Add. There isn't any reason to run dropout removal twice; they should be removed after the first call.

eellison reviewed (commented):
Looks good! mostly just a few nits on the test

desertfire committed:
Update on "[JIT] Remove Dropout during Frozon Optimization"


Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

[ghstack-poisoned]

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1517,3 +1517,25 @@ def test_optimize_freeze_module(self):
+                super(Net, self).__init__()
+                self.dropout = nn.Dropout(0.5)
+
+            def forward(self, x):
```
Could we add a test for aten::feature_dropout too ?

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1517,3 +1517,25 @@ def test_optimize_freeze_module(self):
+        mod = torch.jit.script(Net())
+        script_mod = torch.jit.script(mod)
+        script_mod.eval()
+ 
```
have lint failure here you're going to want to fix.

eellison reviewed (approved):
Cool! LGTM! 

Just as an fyi - in pytorch we usually accept PRs even with comments outstanding with the assumption the non-nit comments will be addressed before landing (or there will be a comment from author on why they're not addressing it)

desertfire committed:
Update on "[JIT] Remove Dropout during Frozon Optimization"


Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Remove Dropout during Frozon Optimization"


Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

[ghstack-poisoned]
